txscript/engine.go

```diff
@@ -555,6 +560,13 @@ func (vm *Engine) checkHashTypeEncoding(hashType SigHashType) error {
 // checkPubKeyEncoding returns whether or not the passed public key adheres to
 // the strict encoding requirements if enabled.
 func (vm *Engine) checkPubKeyEncoding(pubKey []byte) error {
+	if vm.hasFlag(ScriptVerifyWitnessPubKeyType) && vm.witness &&
```

Comment:
To expand on my previous comment, I personally don't think think `vm.witness` should be set at all, but if it is, it should be named `v0witness`.

I think it would be more explicit in general to have the code that wants v0 programs specifically check for it by using `.witnessProgram` and `.witnessVersion`.  In order to simplify the check a helper function such as `vm.isWitessVersionActive(version) bool` could be added.  Then instead of setting `vm.witness = false` as the code currently does, it could just nil out the `.witnesssProgram`.